apm-agent-plugins/apm-servlet-plugin-it/simple-webapp-integration-test/src/test/java/co/elastic/apm/servlet/AbstractTomcatIntegrationTest.java

```diff
@@ -0,0 +1,132 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package co.elastic.apm.servlet;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.networknt.schema.JsonSchema;
+import com.networknt.schema.JsonSchemaFactory;
+import com.networknt.schema.ValidationMessage;
+import okhttp3.HttpUrl;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.mockserver.model.HttpRequest;
+import org.mockserver.model.HttpResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.Network;
+import org.testcontainers.containers.output.Slf4jLogConsumer;
+import org.testcontainers.images.builder.ImageFromDockerfile;
+
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockserver.model.HttpRequest.request;
+
+public class AbstractTomcatIntegrationTest {
+
+    private static final Logger logger = LoggerFactory.getLogger(ServletIntegrationTest.class);
+
+    @ClassRule
+    public static GenericContainer c = new GenericContainer<>(
+        new ImageFromDockerfile()
+            .withDockerfileFromBuilder(builder -> builder
+                .from("tomcat:9")
+                .run("rm -rf /usr/local/tomcat/webapps/*")
+                .copy("ROOT.war", "/usr/local/tomcat/webapps/ROOT.war")
+                // TODO debugging does not work
+                // to get the randomized debugger port, evaluate c.getMappedPort(8000)
+                // then create a remote debug configuration in IntelliJ using this port
+                // Error running 'Debug': Unable to open debugger port (localhost:33049): java.io.IOException "handshake failed - connection prematurally closed"
+                .env("JPDA_ADDRESS", "8000")
+                .env("JPDA_TRANSPORT", "dt_socket")
+                .env("ELASTIC_APM_SERVER_URL", "http://apm-server:1080")
+                .env("ELASTIC_APM_SERVICE_NAME", "servlet-test-app")
+                .env("ELASTIC_APM_IGNORE_URLS", "/apm/*")
+                .expose(8080, 8000)
+                .entryPoint("catalina.sh", "jpda", "run")
+            )
+            // TODO chicken egg problem here: tests require the war to be present, which is built via mvn package, but mvn package executes the tests
```

Comment:
When we were doing the same thing in ZeroTurnaround, we ended up with Groovy-based setup to avoid building test apps at all.
i.e. https://github.com/bsideup/javaagent-boilerplate/tree/master/src/test + https://gist.github.com/bsideup/70091d2f762b8bc7d9255260c0b5ed15 to test Servet-based app.

You can also adjust Maven's lifecycle to execute war packaging before the test.
Also, AFAIK ` maven-failsafe-plugin` will run *after* WAR packaging

Reply:
Using a groovy based setup sounds good. But I also want to be able to test applications deployed in a traditional way.

Reply:
sure, then `maven-failsafe-plugin` is your friend. The only thing is that you loose IDE friendliness of your tests (you can't run them from your IDE)

Ofc there is another option to use Gradle, and use "Delegate build/test actions to Gradle" in your IDE. This way you can have your test app as a dependency and Gradle will automatically build it for you (if not built) before running the test, and you will be able to run the test from IDE as well, but I understand that changing the build system is not what people usually want to do :D Just FYI-ing you about the possibility and advantages